### DISCUSSION TITLE
Bao Hypervisor virtio devices question

### DISCUSSION THREAD

**Discussion by zlaazlaa** [bao-project/bao-hypervisor#156] [General]
I saw dts and config.c of linux, qemu-aarch64 demo, there are 8 virtio device, I don't know why chose this 8 device, not other virtio device. Actually, there are lots of virtio device in the dtb I dumped from qemu, I tried to change some virtio in bao, and the linux demo just can't run, neither when I try to dwindle virtio numbers, could anyone tell me the rule when chose virtio device? I guess 4 device for virtio-console and other 4 device for virtio-net, just guess, XD...
What's more, why the bootargs in dts is 'hvc0', actually I saw 8 hvc in directory /dev, which is /dev/hvc0 ~ /dev/hvc7.

**Comment by josecm** [comment]
The reason for giving access to 8 virtio devices is that these devices reside on the same 4K page. Because Bao's main IO isolation primitive is virtual memory, this is the granularity we have. I don't remember the exact details, but I believe we chose the first 8 devices because when you configure a virtio device in qemu's command line, it assigns the devices to these virtio MMIO "slots" in order. We create two devices in the demo's command line (console and net), this means qemu will assign them the first and second slots. To assign them to linux, we have to assign all the other that reside in the same page. Technically, we could remove the other six from the guest' dtb.

> What's more, why the bootargs in dts is 'hvc0', actually I saw 8 hvc in directory /dev, which is /dev/hvc0 ~ /dev/hvc7.

 I'm guessing the hvc are all the virtio devices Linux sees. And as I explained, probably the console is assigned to the first slot, which is mapped to hvc0.

Again, I don't remember the exact details. But could you explain why this is a problem and exactly what is it out are trying to achieve? Maybe we could be of more assistance.

**Reply by zlaazlaa** [reply]
Thanks very much for your answer, I now know some details about virtio devices.
Acturally I'm trying to run Linux on Bao, I succeed first. But I saw configuration in .platform.regions.place_phys and .platform.regions.phys. I thought this two configuration is not necessary and I delete them, then the Linux can't run on Bao, Which confused me. 
I do some experiment next:
1. Open configuration of .platform.regions.place_phys and .platform.regions.phys, define 8 virtio devices, succeed.
2. Close configuration of .platform.regions.place_phys and .platform.regions.phys, define 8 virtio devices, failed.
3. Close configuration of .platform.regions.place_phys and .platform.regions.phys, NOT define 8 virtio devices, succeed.

So I think it's virtio device matters, I try to understand deeper principles, but got nothing.

My dts on experiment 1 is below:
```c
/dts-v1/;

/ {
	interrupt-parent = <0x8005>;
	#size-cells = <0x02>;
	#address-cells = <0x02>;
	compatible = "linux,dummy-virt";

	psci {
		migrate = <0xc4000005>;
		cpu_on = <0xc4000003>;
		cpu_off = <0x84000002>;
		cpu_suspend = <0xc4000001>;
		method = "hvc";
		compatible = "arm,psci-0.2\0arm,psci";
	};

	memory@60000000 {
		reg = <0x0 0x60000000 0x0 0x40000000>;
		device_type = "memory";
	};

	platform@c000000 {
		interrupt-parent = <0x8005>;
		ranges = <0x00 0x00 0xc000000 0x2000000>;
		#address-cells = <0x01>;
		#size-cells = <0x01>;
		compatible = "qemu,platform\0simple-bus";
	};

	fw-cfg@9020000 {
		dma-coherent;
		reg = <0x00 0x9020000 0x00 0x18>;
		compatible = "qemu,fw-cfg-mmio";
	};

	virtio_mmio@a003000 {
		dma-coherent;
		interrupts = <0x00 0x28 0x01>;
		reg = <0x00 0xa003000 0x00 0x200>;
		compatible = "virtio,mmio";
	};

	virtio_mmio@a003200 {
		dma-coherent;
		interrupts = <0x00 0x29 0x01>;
		reg = <0x00 0xa003200 0x00 0x200>;
		compatible = "virtio,mmio";
	};

	virtio_mmio@a003400 {
		dma-coherent;
		interrupts = <0x00 0x2a 0x01>;
		reg = <0x00 0xa003400 0x00 0x200>;
		compatible = "virtio,mmio";
	};

	virtio_mmio@a003600 {
		dma-coherent;
		interrupts = <0x00 0x2b 0x01>;
		reg = <0x00 0xa003600 0x00 0x200>;
		compatible = "virtio,mmio";
	};

	virtio_mmio@a003800 {
		dma-coherent;
		interrupts = <0x00 0x2c 0x01>;
		reg = <0x00 0xa003800 0x00 0x200>;
		compatible = "virtio,mmio";
	};

	virtio_mmio@a003a00 {
		dma-coherent;
		interrupts = <0x00 0x2d 0x01>;
		reg = <0x00 0xa003a00 0x00 0x200>;
		compatible = "virtio,mmio";
	};

	virtio_mmio@a003c00 {
		dma-coherent;
		interrupts = <0x00 0x2e 0x01>;
		reg = <0x00 0xa003c00 0x00 0x200>;
		compatible = "virtio,mmio";
	};

	virtio_mmio@a003e00 {
		dma-coherent;
		interrupts = <0x00 0x2f 0x01>;
		reg = <0x00 0xa003e00 0x00 0x200>;
		compatible = "virtio,mmio";
	};

	gpio-keys {
		#address-cells = <0x01>;
		#size-cells = <0x00>;
		compatible = "gpio-keys";

		poweroff {
			gpios = <0x8007 0x03 0x00>;
			linux,code = <0x74>;
			label = "GPIO Key Poweroff";
		};
	};

	pl031@9010000 {
		clock-names = "apb_pclk";
		clocks = <0x8000>;
		interrupts = <0x00 0x02 0x04>;
		reg = <0x00 0x9010000 0x00 0x1000>;
		compatible = "arm,pl031\0arm,primecell";
	};

	pl011@9000000 {
		clock-names = "uartclk\0apb_pclk";
		clocks = <0x8000 0x8000>;
		interrupts = <0x00 0x01 0x04>;
		reg = <0x00 0x9000000 0x00 0x1000>;
		compatible = "arm,pl011\0arm,primecell";
	};

	pmu {
		interrupts = <0x01 0x07 0x04>;
		compatible = "arm,armv8-pmuv3";
	};

	intc@8000000 {
		phandle = <0x8005>;
		reg = <0x00 0x8000000 0x00 0x10000 0x00 0x80a0000 0x00 0xf60000>;
		#redistributor-regions = <0x01>;
		compatible = "arm,gic-v3";
		ranges;
		#size-cells = <0x02>;
		#address-cells = <0x02>;
		interrupt-controller;
		#interrupt-cells = <0x03>;

		its@8080000 {
			phandle = <0x8006>;
			reg = <0x00 0x8080000 0x00 0x20000>;
			msi-controller;
			compatible = "arm,gic-v3-its";
		};
	};

	cpus {
		#size-cells = <0x00>;
		#address-cells = <0x01>;

		cpu-map {

			socket0 {

				cluster0 {

					core0 {
						cpu = <0x8004>;
					};

					core1 {
						cpu = <0x8003>;
					};

					core2 {
						cpu = <0x8002>;
					};

					core3 {
						cpu = <0x8001>;
					};
				};
			};
		};

		cpu@0 {
			phandle = <0x8004>;
			reg = <0x00>;
			enable-method = "psci";
			compatible = "arm,cortex-a53";
			device_type = "cpu";
		};

		cpu@1 {
			phandle = <0x8003>;
			reg = <0x01>;
			enable-method = "psci";
			compatible = "arm,cortex-a53";
			device_type = "cpu";
		};

		cpu@2 {
			phandle = <0x8002>;
			reg = <0x02>;
			enable-method = "psci";
			compatible = "arm,cortex-a53";
			device_type = "cpu";
		};

		cpu@3 {
			phandle = <0x8001>;
			reg = <0x03>;
			enable-method = "psci";
			compatible = "arm,cortex-a53";
			device_type = "cpu";
		};
	};

	timer {
		interrupts = <0x01 0x0d 0x04 0x01 0x0e 0x04 0x01 0x0b 0x04 0x01 0x0a 0x04>;
		always-on;
		compatible = "arm,armv8-timer\0arm,armv7-timer";
	};

	apb-pclk {
		phandle = <0x8000>;
		clock-output-names = "clk24mhz";
		clock-frequency = <0x16e3600>;
		#clock-cells = <0x00>;
		compatible = "fixed-clock";
	};

	chosen {
		bootargs = "earlycon console=hvc0;
	};
};
```

My configuration on experiment 1 is below:
```c
struct config config = {

    CONFIG_HEADER

    .vmlist_size = 1,
    .vmlist = {
        {
            .image = {
                .base_addr = 0x60000000,
                .load_addr = VM_IMAGE_OFFSET(linux_image),
                .size = VM_IMAGE_SIZE(linux_image)
            },

            .entry = 0x60000000,

            .platform = {
                .cpu_num = 4,

                .region_num = 1,
                .regions =  (struct vm_mem_region[]) {
                    {
                        .base = 0x60000000,
                        .size = 0x40000000,
                        .place_phys = true,
                        .phys = 0x60000000
                    }
                },

                .dev_num = 4,
                .devs =  (struct vm_dev_region[]) {
                    {
                        /* PL011 */
                        .pa = 0x9000000,
                        .va = 0x9000000,
                        .size = 0x1000,
                        .interrupt_num = 1,
                        .interrupts = (irqid_t[]) {33}
                    },
                    {
                        /* Arch timer interrupt */
                        .interrupt_num = 1,
                        .interrupts = (irqid_t[]) {27}
                    },
                    {
                        /* PL031 */
                        .pa = 0x9010000,
                        .va = 0x9010000,
                        .size = 0x1000,
                        .interrupt_num = 1,
                        .interrupts = (irqid_t[]) {34}
                    },
                    {
                        /* virtio devices */
                        .pa = 0xa003000,   
                        .va = 0xa003000,  
                        .size = 0x1000,
                        .interrupt_num = 8,
                        .interrupts = (irqid_t[]) {72,73,74,75,76,77,78,79}
                    },
                },

                .arch = {
                    .gic = {
                       .gicd_addr = 0x8000000,
                       .gicr_addr = 0x80A0000
                    }
                }
            },
        }
    },
};
```

**Reply by zlaazlaa** [reply]
> The reason for giving access to 8 virtio devices is that these devices reside on the same 4K page. Because Bao's main IO isolation primitive is virtual memory, this is the granularity we have. I don't remember the exact details, but I believe we chose the first 8 devices because when you configure a virtio device in qemu's command line, it assigns the devices to these virtio MMIO "slots" in order. We create two devices in the demo's command line (console and net), this means qemu will assign them the first and second slots. To assign them to linux, we have to assign all the other that reside in the same page. Technically, we could remove the other six from the guest' dtb.
> 
> > What's more, why the bootargs in dts is 'hvc0', actually I saw 8 hvc in directory /dev, which is /dev/hvc0 ~ /dev/hvc7.
> 
> I'm guessing the hvc are all the virtio devices Linux sees. And as I explained, probably the console is assigned to the first slot, which is mapped to hvc0.
> 
> Again, I don't remember the exact details. But could you explain why this is a problem and exactly what is it out are trying to achieve? Maybe we could be of more assistance.

In my opinion, if we open configuration of .platform.regions.place_phys and .platform.regions.phys, Bao will put image in the physical address we assigned, and if we close these two configuration, Bao will automatically find idle physical address on qemu. So these two configuration are not matter, we can close them if we like. Is my opinion correct?

**Comment by zlaazlaa** [comment]
What's more, even change the value of .phys also matters the result, only `.phys = 0x60000000` can run successfully. Why 0x60000000? I'm confusing. 🥺😭

**Comment by josecm** [comment]
> But I saw configuration in .platform.regions.place_phys and .platform.regions.phys. I thought this two configuration is not necessary and I delete them, then the Linux can't run on Bao, Which confused me.

The answer to this is quite straightforward. Virtio devices are essentially DMA devices. However, these are not connected to an IOMMU (or if there is a virtio-iommu, Bao does not support it). Therefore, we need to configure identity mappings for the guest so that they work correctly. If you remove the identity mapping (i.e., setting the place phys options with phys == base) , Bao places the guest in a different physical address, and devices will be given the wrong addresses by the guest.

> Why 0x60000000?

This is just a choice for the configuration. We could chose another physical address to place the guest. Just change it in the device tree and make sure in the guest region configuration .base is the same as .phys

**Reply by zlaazlaa** [reply]
Thanks! So this special situation is caused by virtio, this has solved my long-standing doubts, Thanks again!
